Automattic/wp-calypso
# State: Consolidate post tracking into query managers

This pull request seeks to consolidate post tracking into the `posts.queries` state. Currently, post data is duplicated between both `posts.items` and `posts.queries` (`QueryManager` instances). Not only is this a syncing problem, but is also memory wasteful. In order to move forward with state persistence of query data, we should seek to be more efficient with the storage of posts, which is the intent of these changes.

__Implementation notes:__

To continue to allow post lookup by global ID, the `posts.items` state remains but has been transformed in purpose to map to a "path" tuple of site ID, post ID pair. While it's true this could have been achieved in a selector instead of storing this in state, it would be very inefficient to look up a post across all sites stored in `posts.queries` to try to find the item matching the global ID.

__Testing instructions:__

There are no visual effects of these changes. Verify that there are no regressions in using either the [custom post types list screen](http://calypso.localhost:3000/types/post) or the [post editor "Add existing content" / "Parent page" selectors](http://calypso.localhost:3000/post).

Ensure Mocha tests pass:

```
npm run test-client
```

__Follow-up tasks:__

- Persist query state
- Implement memoization for `normalizePost` (we memoize the selectors, but this is busted more frequently than desirable and we can probably be more efficient)

Test live: https://calypso.live/?branch=update/state-posts-dedupe
